### PR TITLE
Port CommentTrackingXmlPullParser from tools/base/resource-repository

### PR DESCRIPTION
**Pull request by kevinzheng-ap**
Part of https://github.com/cashapp/paparazzi/issues/524

cc @jrodbx

### CODE REVIEW

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+      setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, true)
+    } catch (e: XmlPullParserException) {
+      throw Error(e) // KXmlParser is guaranteed to support FEATURE_PROCESS_NAMESPACES.
+    }
+  }
+
+  /**
+   * Returns the name of the current "attr" group, e.g. "Button Styles" group for "buttonStyleSmall" "attr" tag.
+   */
+  fun getAttrGroupComment(): String? = attrGroupCommentStack[attrGroupCommentStack.size - 1]
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt:61]: nit: can also be converted to a val

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+  private fun processToken(token: Int) {
+    when (token) {
+      XmlPullParser.START_TAG -> {
+        if (tagEncounteredAfterComment) {
+          lastComment = null
+        }
+        tagEncounteredAfterComment = true
+        // Duplicate the last element in attrGroupCommentStack.
+        attrGroupCommentStack.add(attrGroupCommentStack[attrGroupCommentStack.size - 1])
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt:83]:
```suggestion
        attrGroupCommentStack += attrGroupCommentStack.last()
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        assert(attrGroupCommentStack.size == depth + 1)
+        if (TAG_EAT_COMMENT == name && prefix == null) {
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt:85]:
```suggestion

        if (TAG_EAT_COMMENT == name && prefix == null) {
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+          //     <!-- Text styles -->
+          //     <!-- =========== -->
+          //     <eat-comment/>
+          // These section headers are always immediately followed by an <eat-comment>. Not all <eat-comment/> sections are
+          // actually attribute headers, some are comments. We identify these by looking at the line length; category comments
+          // are short, and descriptive comments are longer.
+          if (lastComment != null && lastComment!!.length <= ATTR_GROUP_MAX_CHARACTERS && !lastComment!!.startsWith(
+              "TODO:"
+            )
+          ) {
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt:98]:
```suggestion
          if (lastComment != null && lastComment!!.length <= ATTR_GROUP_MAX_CHARACTERS && !lastComment!!.startsWith("TODO:")) {
```

or basically, any way that the line breaks can be setup without "TODO:" on its own line?

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            var attrGroupComment = lastComment!!
+            if (attrGroupComment.endsWith(".")) {
+              attrGroupComment = attrGroupComment.substring(
+                0,
+                attrGroupComment.length - 1
+              ) // Strip the trailing period.
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt:104]:
```suggestion
              // Strip the trailing period.
              attrGroupComment = attrGroupComment.substring(0,attrGroupComment.length - 1)
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            }
+            // Replace the second to last element in attrGroupCommentStack.
+            attrGroupCommentStack[attrGroupCommentStack.size - 2] = attrGroupComment
+          }
+        }
+      }
+
+      XmlPullParser.END_TAG -> {
+        lastComment = null
+        attrGroupCommentStack.removeAt(attrGroupCommentStack.size - 1)
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt:114]:
```suggestion
        attrGroupCommentStack.removeLast()
```

File: paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    lastComment = null
+    attrGroupCommentStack.clear()
+    attrGroupCommentStack.add(null)
+  }
+
+  companion object {
+    // Used for parsing group of attributes, used heuristically to skip long comments before <eat-comment/>.
+    private const val ATTR_GROUP_MAX_CHARACTERS = 40
+
+    private fun isEmptyOrAsciiArt(commentText: String): Boolean = commentText.isEmpty() || commentText[0] == '*' || commentText[0] == '='
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/main/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParser.kt:147]:
```suggestion
    private fun isEmptyOrAsciiArt(commentText: String): Boolean = 
      commentText.isEmpty() || commentText[0] == '*' || commentText[0] == '='
```

File: paparazzi/paparazzi/src/test/java/app/cash/paparazzi/internal/resources/ValueResourceXmlParserTest.kt [hunk 8]
```diff
@@ -3,63 +3,63 @@ package app.cash.paparazzi.internal.resources
...
 class ValueResourceXmlParserTest {
   @Test
   fun test() {
     StringReader(SIMPLE_LAYOUT).use {
       val parser = ValueResourceXmlParser().apply {
-        setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, true)
         setInput(it)
       }
 
+      parser.nextToken()
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/test/java/app/cash/paparazzi/internal/resources/ValueResourceXmlParserTest.kt:16]: necessary?  (why was this not needed before on the left?)

File: paparazzi/paparazzi/src/test/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParserTest.kt [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+class CommentTrackingXmlPullParserTest {
+  @Test
+  fun test() {
+    StringReader(SIMPLE_RES).use {
+      val parser = CommentTrackingXmlPullParser().apply {
+        setInput(it)
+      }
+
+      // <resources>
+      parser.nextToken()
```
Comment (jrodbx) [inline paparazzi/paparazzi/src/test/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParserTest.kt:16]: similar to question in ValueResourceXmlParser, is this needed?

**Review by jrodbx — APPROVED** [review]

File: paparazzi/paparazzi/src/test/java/app/cash/paparazzi/internal/resources/ValueResourceXmlParserTest.kt (hunk 8 above)
Comment (kevinzheng-ap) [reply paparazzi/paparazzi/src/test/java/app/cash/paparazzi/internal/resources/ValueResourceXmlParserTest.kt:16]: As discussed , call `next()` will skip ENTITY_REF until reach next valid token, however, `nextToken()` will do one step, regardless of what the token is. But `next()` is disabled in `CommentTrackingXmlPullParser`, so we have to switch to `nextToken()`

File: paparazzi/paparazzi/src/test/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParserTest.kt (hunk 9 above)
Comment (kevinzheng-ap) [reply paparazzi/paparazzi/src/test/java/app/cash/paparazzi/internal/resources/CommentTrackingXmlPullParserTest.kt:16]: ditto

**Pull request merged**